Sources/GRPCCore/Call/Server/RPCRouter.swift

```diff
@@ -0,0 +1,135 @@
+      method: MethodDescriptor,
+      deserializer: some MessageDeserializer<Input>,
+      serializer: some MessageSerializer<Output>,
+      handler: @Sendable @escaping (
+        _ request: ServerRequest.Stream<Input>
+      ) async throws -> ServerResponse.Stream<Output>
+    ) {
+      self._fn = { stream, interceptors in
+        await ServerRPCExecutor.execute(
+          stream: stream,
+          deserializer: deserializer,
+          serializer: serializer,
+          interceptors: interceptors,
+          handler: handler
+        )
+      }
+    }
+
+    @inlinable
+    func handle(
+      stream: RPCStream<RPCAsyncSequence<RPCRequestPart>, RPCWriter<RPCResponsePart>.Closable>,
+      interceptors: [any ServerInterceptor]
+    ) async {
+      await self._fn(stream, interceptors)
+    }
+  }
+
+  @usableFromInline
+  private(set) var handlers: [MethodDescriptor: RPCHandler]
+
+  /// Creates a new router with no methods registered.
+  public init() {
+    self.handlers = [:]
+  }
+
+  /// Returns all descriptors known to the router in an undefined order.
+  public var methods: [MethodDescriptor] {
+    Array(self.handlers.keys)
+  }
+
+  /// Returns the number of methods registered with the router.
+  public var count: Int {
+    self.handlers.count
+  }
+
+  /// Returns whether a handler exists for a given method.
+  ///
+  /// - Parameter descriptor: A descriptor of the method.
+  /// - Returns: Whether a handler exists for the method.
+  public func hasHandler(forMethod descriptor: MethodDescriptor) -> Bool {
+    return self.handlers.keys.contains(descriptor)
+  }
+
+  /// Registers a handler with the router.
+  ///
+  /// - Note: if a handler already exists for a given method then it will be replaced.
+  ///
+  /// - Parameters:
+  ///   - descriptor: A descriptor for the method to register a handler for.
+  ///   - deserializer: A deserializer to deserialize input messages received from the client.
+  ///   - serializer: A serializer to serialize output messages to send to the client.
+  ///   - handler: The function which handles the request and returns a response.
+  @inlinable
+  public mutating func registerHandler<Input: Sendable, Output: Sendable>(
+    forMethod descriptor: MethodDescriptor,
+    deserializer: some MessageDeserializer<Input>,
+    serializer: some MessageSerializer<Output>,
+    handler: @Sendable @escaping (
+      _ request: ServerRequest.Stream<Input>
+    ) async throws -> ServerResponse.Stream<Output>
+  ) {
+    self.handlers[descriptor] = RPCHandler(
+      method: descriptor,
+      deserializer: deserializer,
+      serializer: serializer,
+      handler: handler
+    )
+  }
+
+  /// Removes any handler registered for the specified method.
+  ///
+  /// - Parameter descriptor: A descriptor of the method to remove a handler for.
+  /// - Returns: Whether a handler was removed.
+  @discardableResult
+  public mutating func removeHandler(forMethod descriptor: MethodDescriptor) -> Bool {
```

Comment:
Same here. Does this need to be public?

Reply:
If you don't want to offer all methods in a service then this is one way to remove the ones you don't want (the other being to reimplement the service protocol).

Reply:
I see. I am wondering if we should expose this publicly right away. Do we have concrete use-cases in mind where want to do this or did we get a feature request for this before?

Reply:
One concrete use case is the interoperability tests. One test uses the generated client to call a method which isn't implemented on the server. At the moment we achieve this by patching the generated code.

Reply:
I also think having these methods is very low burden: they're unlikely to change and they still make sense if we change the internals of the router.